### ISSUE TITLE
`redundant_guards` doesn't take `ref` into account

### ISSUE THREAD

**Issue by Serial-ATA** [rust-lang/rust-clippy#11465]
### Summary

If I have an enum variant storing an array, whose value is compared to byte strings, the comparison must happen in a guard.

For example, I'd do:

`Foo(ref arr) if arr == b"bar"`

Where `arr` is `[u8; 3]`.

The lint will incorrectly assume that `b"bar"` can simply be moved into the pattern, which is not the case, as its type would be `&[u8; 3]`.

### Lint Name

redundant_guards

### Reproducer

I tried this code:

```rust
enum Test {
    Foo([u8; 3])
}

let foo = Test::Foo([98, 97, 114]);

match foo {
    Test::Foo(ref arr) if arr == b"bar" => {}
    _ => {}
}
```

I saw this happen:

```
warning: redundant guard
  --> src/main.rs:11:27
   |
11 |     Test::Foo(ref arr) if arr == b"bar" => {}
   |                           ^^^^^^^^^^^^^
   |
   = help: for further information visit https://rust-lang.github.io/rust-clippy/master/index.html#redundant_guards
   = note: `#[warn(clippy::redundant_guards)]` on by default
help: try
   |
11 -     Test::Foo(ref arr) if arr == b"bar" => {}
11 +     Test::Foo(b"bar") => {}
   |
```

I expected to see this happen:

No warning

### Version

```text
rustc 1.74.0-nightly (ec08a0337 2023-09-04)
binary: rustc
commit-hash: ec08a0337f3556212525dbf1d3b41e19bdf27621
commit-date: 2023-09-04
host: x86_64-unknown-linux-gnu
release: 1.74.0-nightly
LLVM version: 17.0.0
```

### Additional Labels

@rustbot label +I-suggestion-causes-error
